diff --git a/src/fellow_log.c b/src/fellow_log.c
index b2cbb8b..3cf2055 100644
--- a/src/fellow_log.c
+++ b/src/fellow_log.c
@@ -5470,6 +5470,12 @@ fellow_logs_iter(const struct flics *flics, struct flivs *flivs,
 		}
 
 		fellow_logs_iter_block(flics, flivs, logblk);
+		if (flics->ffd->phase == FP_FINI && flivs->tofree != NULL) {
+			regionlist_stk_flush(tofree);
+			if (regionlist_used(flivs->tofree))
+				regionlist_drain(flivs->tofree,
+				    flics->ffd->dskbuddy);
+		}
 
 		off = logblk->prev_off;
 	}
