diff --git a/src/fellow_log.c b/src/fellow_log.c
index 3cf2055..98a7eca 100644
--- a/src/fellow_log.c
+++ b/src/fellow_log.c
@@ -6409,8 +6409,6 @@ fellow_logwatcher_rewrite_work(void *priv)
 
 	AZ(pthread_mutex_lock(&ffd->logmtx));
 	ffd->rewrite_state = FLW_S_NOREWRITE;
-	if (ffd->watcher_running == 2)
-		ffd->watcher_running = 1;
 	AZ(pthread_cond_broadcast(&ffd->watcher_cond));
 	AZ(pthread_mutex_unlock(&ffd->logmtx));
 	return (NULL);
@@ -6561,11 +6559,6 @@ fellow_logwatcher_fini(struct fellow_fd *ffd)
 	CHECK_OBJ_NOTNULL(ffd, FELLOW_FD_MAGIC);
 
 	AZ(pthread_mutex_lock(&ffd->logmtx));
-	ffd->watcher_running = 2;
-	AZ(pthread_cond_signal(&ffd->watcher_cond));
-	while (ffd->watcher_running == 2)
-		AZ(pthread_cond_wait(&ffd->watcher_cond, &ffd->logmtx));
-	assert(ffd->watcher_running == 1);
 	ffd->watcher_running = 0;
 	AZ(pthread_cond_signal(&ffd->watcher_cond));
 	AZ(pthread_mutex_unlock(&ffd->logmtx));
