diff --git a/src/fellow_log.c b/src/fellow_log.c
index 70ff8d5..75730c0 100644
--- a/src/fellow_log.c
+++ b/src/fellow_log.c
@@ -738,6 +738,7 @@ struct fellow_fd {
 	pthread_cond_t			watcher_cond;
 	pthread_t			watcher_thread;
 	unsigned			watcher_running;
+	unsigned			watcher_shutdown;
 	enum fellow_watcher_state	watcher_state;
 	enum fellow_rewrite		rewrite_state;
 
@@ -6549,8 +6550,11 @@ fellow_logwatcher_thread(void *priv)
 	while (ffd->watcher_running) {
 		if (ffd->rewrite_state == FLW_S_REWRITING)
 			need = FLW_MAYFLUSH;
-		else
+		else {
 			need = fellow_logwatcher_need(ffd, NULL);
+			if (ffd->watcher_shutdown && need == FLW_NEED_REWRITE)
+				need = FLW_MAYFLUSH;
+		}
 
 		switch (need) {
 		case FLW_NEED_NONE:
@@ -6577,7 +6581,7 @@ fellow_logwatcher_thread(void *priv)
 			assert(ffd->rewrite_state == FLW_S_NOREWRITE);
 			ffd->watcher_state = FLW_S_FLUSHING;
 			fellow_logwatcher_flush(ffd, 0);
-			if (! ffd->watcher_running)
+			if (! ffd->watcher_running || ffd->watcher_shutdown)
 				continue;
 			ffd->rewrite_state = FLW_S_REWRITING;
 			AZ(pthread_mutex_unlock(&ffd->logmtx));
@@ -6673,6 +6677,18 @@ fellow_logwatcher_fini(struct fellow_fd *ffd)
 	AZ(pthread_cond_destroy(&ffd->watcher_cond));
 }
 
+void
+fellow_logwatcher_shutdown(struct fellow_fd *ffd)
+{
+	CHECK_OBJ_NOTNULL(ffd, FELLOW_FD_MAGIC);
+
+	AZ(pthread_mutex_lock(&ffd->logmtx));
+	ffd->watcher_shutdown = 1;
+	if (ffd->watcher_running)
+		AZ(pthread_cond_signal(&ffd->watcher_cond));
+	AZ(pthread_mutex_unlock(&ffd->logmtx));
+}
+
 static void *
 fellow_space_thread(void *priv)
 {
diff --git a/src/fellow_log_storage.h b/src/fellow_log_storage.h
index f7ecc34..7f332d6 100644
--- a/src/fellow_log_storage.h
+++ b/src/fellow_log_storage.h
@@ -249,6 +249,7 @@ fellow_log_open(struct fellow_fd *ffd,
 void fellow_log_flush(struct fellow_fd *ffd);
 
 void fellow_log_close(struct fellow_fd **ffdp);
+void fellow_logwatcher_shutdown(struct fellow_fd *ffd);
 
 int fellow_log_ban(struct fellow_fd *ffd, uint8_t op,
     const uint8_t *bans, unsigned len, vtim_real time,
diff --git a/src/fellow_storage.c b/src/fellow_storage.c
index 0ad9604..0bcce2c 100644
--- a/src/fellow_storage.c
+++ b/src/fellow_storage.c
@@ -2930,6 +2930,7 @@ sfe_close_warn(struct stvfe *stvfe)
 	if (stvfe->oev)
 		ObjUnsubscribeEvents(&stvfe->oev);
 	fellow_log_flush(stvfe->ffd);
+	fellow_logwatcher_shutdown(stvfe->ffd);
 	stvfe->shutdown = 1;
 	fellow_cache_shutdown(stvfe->fc);
 	buddy_kick(fellow_dskbuddy(stvfe->ffd));
